File: app/src/main/java/com/github/kongpf8848/animation/splash/SplashAnimation.java
package com.github.kongpf8848.animation.splash;

import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.Transformation;
import android.widget.FrameLayout;
import android.widget.RelativeLayout;

import com.github.kongpf8848.animation.BaseApplication;
import com.kongpf.commonhelper.ScreenHelper;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.Random;

public class SplashAnimation extends Animation {

    private Random random = new Random();

    private int textStart=0;
    private int textEnd=0;
    private int picStart=0;
    private int picEnd=0;

    private SplashView mSplashView;
    public ArrayList<TopView> topViews = new ArrayList<>();

    public static class TopView
    {
        int margin = ScreenHelper.dp2px(BaseApplication.getContext(), 10.0F);
        public View view;
        public float startAlpha;
        public int topMargin;
        FrameLayout.LayoutParams layoutParams;

        public TopView(View view, float alpha)
        {
            this.view = view;
            this.startAlpha = alpha;
            this.layoutParams = ((FrameLayout.LayoutParams)view.getLayoutParams());
            this.topMargin = this.layoutParams.topMargin;
        }

        public void doAnimation(float alpha)
        {
            float half = 0.5F;
            float temp = alpha - this.startAlpha;
            if (temp > half)
            {
                return;
            }

            float result = temp * 2.0F;
            this.view.setAlpha(result);
            this.layoutParams.topMargin = ((int)((1.0F - result) * margin + this.topMargin));

            Log.d("JACK8","topmargin:"+this.layoutParams.topMargin+",alpha:"+alpha);
            this.view.setLayoutParams(this.layoutParams);

        }
    }


    public SplashAnimation(SplashView splashView)
    {
        this.mSplashView=splashView;
        int screenWidth = ScreenHelper.getScreenWidth(BaseApplication.applicationContext);

        int textWidth = this.mSplashView.text.getMeasuredWidth();
        this.textStart = -(textWidth / 2);
        this.textEnd = (screenWidth - textWidth) / 2;

        int picWidth = this.mSplashView.pic.getMeasuredWidth();
        this.picStart = -(picWidth / 2);
        this.picEnd = (screenWidth - picWidth) / 2;
    }

    /**
     * 做透明度和位移动画
     * @param view
     * @param alpha
     * @param start
     * @param end
     */
    private void alphaAnimation1(View view,float alpha,int start,int end)
    {
        if (view== null) {
            return;
        }
        view.setAlpha(alpha);
        RelativeLayout.LayoutParams layoutParams = (RelativeLayout.LayoutParams)view.getLayoutParams();
        if (layoutParams == null) {
            layoutParams = new RelativeLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        }
        layoutParams.rightMargin = ((int)(start + alpha * (end-start)));
        Log.d("JACK8", "alphaAnimation called with:view="+view+", alpha = [" + alpha + "]"+",rightMargin:"+layoutParams.rightMargin);
        view.setLayoutParams(layoutParams);
    }

    /**
     * 缩放动画
     * @param alpha
     */
    private void scalaAnimation(float alpha)
    {
        if ((this.mSplashView == null) || (this.mSplashView.certificate == null)) return;
        this.mSplashView.certificate.setAlpha(alpha);
        this.mSplashView.certificate.setScaleX(2.0F - alpha);
        this.mSplashView.certificate.setScaleY(2.0F - alpha);
    }

    /**
     * 透明度和位移动画
     * @param alpha
     */
    private void alphaAnimation2(float alpha)
    {
        if ((this.mSplashView.views == null) || (this.mSplashView.views.isEmpty())) return;

        if ((alpha > 0.5F * this.topViews.size()/ this.mSplashView.views.size())
         && (this.topViews.size() < this.mSplashView.views.size()))
        {
            Log.d("JACK8","alphaAnimation2:"+alpha);
            addTopView(alpha);
        }

        for (TopView view : this.topViews) {
            view.doAnimation(alpha);
        }
    }



    private boolean inTopView(View view)
    {
        if (this.topViews.isEmpty()) {
            return false;
        }
        for (TopView topView : this.topViews) {
            if (topView.view == view) {
                return true;
            }
        }
        return false;
    }

    private void addTopView(float alpha)
    {
        int bound = this.mSplashView.views.size();
        int j=0;
        for(j=random.nextInt(bound); inTopView(this.mSplashView.views.get(j)); j = random.nextInt(bound));
        TopView topView = new TopView(this.mSplashView.views.get(j), alpha);
        this.topViews.add(topView);
    }



    @Override
    protected void applyTransformation(float interpolatedTime, Transformation t)
    {
        if (interpolatedTime <=0.25F)
        {
            alphaAnimation1(mSplashView.text,interpolatedTime * 4.0F,textStart,textEnd);
        }
        if ((interpolatedTime > 0.125F) && (interpolatedTime <= 0.375F)) {
            alphaAnimation1(mSplashView.pic,4.0F * (interpolatedTime - 0.125F),picStart,picEnd);
        }
        if (interpolatedTime >0.375F)
        {
            scalaAnimation(1.6F * (interpolatedTime - 0.375F));
            alphaAnimation2(1.6F * (interpolatedTime - 0.375F));

        }
    }
}

File: app/src/main/java/com/github/kongpf8848/animation/splash/SplashAnimation.kt
package com.github.kongpf8848.animation.splash

import android.animation.IntEvaluator
import android.util.Log
import android.view.View
import android.view.animation.Animation
import android.view.animation.Transformation
import android.widget.FrameLayout
import android.widget.RelativeLayout
import com.github.kongpf8848.animation.TKApplication
import com.kongpf.commonhelper.ScreenHelper
import java.util.*

class SplashAnimation(private val mSplashView: SplashView) : Animation() {

    private val random = Random()
    private var textStart = 0
    private var textEnd = 0
    private var picStart = 0
    private var picEnd = 0

    var topViews = mutableListOf<TopView>()

    init {
        val screenWidth = ScreenHelper.getScreenWidth(TKApplication.instance)
        val textWidth = mSplashView.text?.measuredWidth?:0
        textStart = -(textWidth / 2)
        textEnd = (screenWidth - textWidth) / 2
        val picWidth = mSplashView.pic?.measuredWidth?:0
        picStart = -(picWidth / 2)
        picEnd = (screenWidth - picWidth) / 2
    }

    class TopView(var view: View, var startAlpha: Float) {
        val margin = ScreenHelper.dp2px(TKApplication.instance, 10.0f)
        val layoutParams: FrameLayout.LayoutParams = view.layoutParams as FrameLayout.LayoutParams
        val topMargin=layoutParams.topMargin

        fun doAnimation(alpha: Float) {
            val half = 0.5f
            val temp = alpha - startAlpha
            if (temp > half) {
                return
            }
            val fraction = temp * 2.0f
            view.alpha = fraction
            layoutParams.topMargin = ((1.0f - fraction) * margin + topMargin).toInt()
            view.layoutParams = layoutParams
        }

    }

    /**
     * 做透明度和位移动画
     * @param view
     * @param fraction
     * @param start
     * @param end
     */
    private fun alphaAnimation1(view: View?, fraction: Float, start: Int, end: Int) {
        if (view == null) {
            return
        }
        view.alpha = fraction
        val layoutParams = view.layoutParams as RelativeLayout.LayoutParams
        layoutParams.rightMargin = IntEvaluator().evaluate(fraction,start,end)
        view.layoutParams = layoutParams
        Log.d("JACK8", "alphaAnimation called with:view=" + view + ", fraction = [" + fraction + "]" + ",rightMargin:" + layoutParams.rightMargin)

    }

    /**
     * 透明度和位移动画
     * @param alpha
     */
    private fun alphaAnimation2(alpha: Float) {
        if (mSplashView.views.isNullOrEmpty()) return
        if (alpha > 0.5f * topViews.size / mSplashView.views!!.size
                && (topViews.size < mSplashView.views!!.size)) {
            Log.d("JACK8", "alphaAnimation2:$alpha")
            addTopView(alpha)
        }
        topViews.forEach {
            it.doAnimation(alpha)
        }
    }

    /**
     * 缩放动画
     * @param fraction
     */
    private fun scalaAnimation(fraction: Float) {
        mSplashView.certificate?.apply {
            alpha = fraction
            scaleX = 2.0f - fraction
            scaleY = 2.0f - fraction
        }

    }

    private fun inTopView(view: View): Boolean {
       topViews.forEach {
           if (it.view === view) {
               return true
           }
       }
        return false
    }

    private fun addTopView(alpha: Float) {
        val count = mSplashView.views!!.size
        var j = 0
        j = random.nextInt(count)
        while (inTopView(mSplashView.views!![j])) {
            j = random.nextInt(count)
        }
        val topView = TopView(mSplashView.views!![j], alpha)
        topViews.add(topView)
    }

    override fun applyTransformation(interpolatedTime: Float, t: Transformation) {
        if (interpolatedTime <= 0.25f) {
            alphaAnimation1(mSplashView.text, interpolatedTime * 4.0f, textStart, textEnd)
        }
        if (interpolatedTime > 0.125f && interpolatedTime <= 0.375f) {
            alphaAnimation1(mSplashView.pic, 4.0f * (interpolatedTime - 0.125f), picStart, picEnd)
        }
        if (interpolatedTime > 0.375f) {
            scalaAnimation(1.6f * (interpolatedTime - 0.375f))
            alphaAnimation2(1.6f * (interpolatedTime - 0.375f))
        }
    }


}